Extend custom content renderer API

danigm opened the pull request:
## Summary

Extend the custom content renderer javascript API to be able to do advanced content query and add a random filter to get the content randomly ordered.

## Testing checklist

- [*] Contributor has fully tested the PR manually

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

danigm committed:
Add random filter to fetchContentCollection

This patch adds an extra option to the fetchContentCollection method to
be able to do advanced queries to the API using the window.kolibri
object in the custom content renderer.

This random filter could be used to get random content, using
`order_by('?')` [1].

The filter is a CharFilter to avoid the cache, to get randomly ordered
content each query should have a different `random` parameter.

The page and pageSize parameters are also passed to the
fetchContentCollection

This patch also limit the content returned by fetchContentCollection to
the current channel to not allow to get content outside of the scope and
adds another filter to exclude topics from the list of returned nodes.

[1] https://docs.djangoproject.com/en/3.2/ref/models/querysets/#order-by

rtibbles commented on kolibri/plugins/learn/assets/src/views/ChannelRenderer/CustomContentRenderer.vue:
```diff
@@ -117,6 +117,7 @@
           getParams: {
             ids: options.ids,
             parent: options.parent === 'self' ? this.topic.id : options.parent,
+            ...options.extra,
```
I'd rather have these continue to be explicit to keep the API documented, otherwise, we might look for where the `random` filter in the FilterSet is being used and realize we don't have any code references to it anywhere, and then delete it.

manuq replied:
Yes I agree. I see the way you @danigm found to pass extra parameters as a fantastic starting point, but it's a shortcut. All options should be explicit and well documented. Not directly accessing the backend from the frontend.

From the frontend, the API could be designed like this for fetching 5 random results using pagination:

```
options: {
  randomize_results: true,
  page_size: 5,
  page: 1,
}
```

Or add a separate API method to fetch N random content nodes, which is all we need. But that would only fit our current needs.

rtibbles commented on packages/hashi/src/kolibri.js:
```diff
@@ -94,6 +94,7 @@ export default class Kolibri extends BaseShim {
        * @param {string[]} options.ids - an array of ids to filter by
        * @param {number} [options.page=1] - which page to return from the result set
        * @param {number} [options.pageSize=50] - the page size for pagination
+       * @param {Object} options.extra - extra parameters to pass to the fetchCollection call
```
Likewise here.

manuq replied:
Yes I agree, all should be explicit and documented. No "extra" parameters.

rtibbles commented on kolibri/core/content/api.py:
```diff
@@ -176,6 +179,17 @@ class Meta:
+                      by the cache system, so same value could return same order
+        :return: content nodes randomly ordered
+        """
+        return queryset.order_by("?")
```
I assume there's no need of a guarantee of persistent randomization between different clients for the same 'seed' value here?

manuq replied:
No, there is no need to fetch the same results passing the same seed value.

danigm commented:
Thanks for the review, I've updated the PR to a more specific solution instead of the generic one. I left the random filter, but also added another one to also exclude topic contents.

In this solution I've also added the pagination parameters to the query, because it was not used.

I've also limited the `fetchColletion` queries in this API to the channel nodes, to avoid leaking outside the sandbox. But I'm not sure if this should be a restriction or if it could be an optional parameter something like `channel: 'self'`.

manuq commented on packages/hashi/src/kolibri.js:
```diff
@@ -94,6 +94,9 @@ export default class Kolibri extends BaseShim {
        * @param {number} [options.page=1] - which page to return from the result set
        * @param {number} [options.pageSize=50] - the page size for pagination
+       * @param {string} options.random - set to a random string to get the result randomly ordered
+       * @param {boolean} options.only_content - set to true to skip topics
```
All options so far are cameCase, so I would expect the same here.

manuq commented on packages/hashi/src/kolibri.js:
```diff
@@ -94,6 +94,9 @@ export default class Kolibri extends BaseShim {
        * @param {string[]} options.ids - an array of ids to filter by
        * @param {number} [options.page=1] - which page to return from the result set
        * @param {number} [options.pageSize=50] - the page size for pagination
+       * @param {string} options.random - set to a random string to get the result randomly ordered
```
From the user of the API, it would be better to say "randomize: true" than passing a random string. Or is this option the random seed?

danigm replied:
It's a random "seed". The API is caching every request, so we can't use a boolean here because in that case we'll get the same ordering on every request with the same parameters.

danigm commented:
I'm  closing this PR for now. We'll work on this on our fork and we can create the PR once it's ready for review.
